packages/dds/tree/docs/wip/Container and Application Schema.md

```diff
@@ -0,0 +1,194 @@
+
+# Example/Proposed Design
+
+This design shows an easy to implement MVP, and is not intended to have document or API compatibility with what we do longer term.
+It minimally implements the requirements from "What options to support" while being extensible in the future.
+It can be subseted for use in earlier milestones, and could be extended for later ones, see "Schedule" below.
+
+This design is intentionally minimal while still being useful.
+Features not needed for minimal strongly typed application use or the JSON and XML domain schema are not included.
+
+## Schema Representation
+
+Include a simple non-contextual schema system, usable as a MVP for both `container schema` and `application schema`.
+
+We can start with:
+
+```typescript
+enum Multiplicity {
+    Value,
+    Optional,
+    Sequence,
+}
+
+enum Value {
+    Nothing,
+    Number,
+    String,
+    Boolean,
+    Serializable,
+}
+
+interface FieldContent {
+    multiplicity: Multiplicity;
+    types?: Type[];
+}
+
+interface Field {
+    content: FieldContent;
+    name: string;
+}
+
+interface Type {
+    name: string;
+    fields: Field[];
+    extraFields: FieldContent;
+    value: Value;
+}
+```
+
+These `Types` can be added as `container schema` as part of an edit op, which is considered conflicted if it tries to add a type that has a conflicting `name` and is not equal to the existing one.
+
+This intentionally does not support any concept of versioning or name-spacing: users of it can include versions and namespaces in the `name` if they want, or features for them could be added later.
+
+While a schema must be added to the document as a container schema to use the type
+(otherwise adding a new type could break existing data which might not even be downloaded on the current client),
+it's possible to add a schema that is compatible with all possible data.
+Applications which wish to rely entirely on schema-on-read for some or all of their data can use this pattern for all `container schema` and only use their actual developer authored schema as `application schema`:
+
+```typescript
+const anyField: FieldContent = { multiplicity: Multiplicity.sequence };
+function makeAnyType(name: string): Type {
+    return { name, extraFields: anyField };
+}
+```
+
+## Use as `Container Schema`
+
+-   Document load: The application can check that their application schema match container schema, and check that the root has supported content.
+-   Writers: When inserting new content, and updating modified nodes, share-tree can check them against the schema in the document.
+    If the schema is missing, it must be added as part of the edit: for this the editing API needs to take in types which can have their desired container schema queried.
```

Comment:
Yes, that is accurate.

There is the caveat that we could support ops that modify/replace a schema (either in ways were existing data will still be valid or including a process to update all existing instances).

Also someone with a partial checkout would want to download any relevant schema (or just all of them): since we want to require clients to only submit edits which are valid when they submit the edit, they need to know that there is not a schema with the same name, but different content, when they go to add a new schema. Some apps might be able to do this without checking for a schema in the document (ex: an app that centrally manages their schema and never updates schema, or one that includes hashes of schema in their name), but in general its not safe.